src/internal/middleware/errors/interceptor.go

```diff
@@ -5,28 +5,55 @@ import (
 	"errors"
 
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
 // UnaryServerInterceptor translates errors for unary RPCs
 func UnaryServerInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
 	res, err := handler(ctx, req)
-	return res, unwrapGRPC(err)
+	return res, errorForGRPC(err)
 }
 
 // StreamServerInterceptor translates errors for streaming RPCs
 func StreamServerInterceptor(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
 	err := handler(srv, stream)
-	return unwrapGRPC(err)
+	return errorForGRPC(err)
 }
 
-func unwrapGRPC(err error) error {
+func errorForGRPC(err error) error {
+	if _, ok := status.FromError(err); ok {
+		// handles nil as well
+		return err //nolint:wrapcheck
+	}
+	code := codes.Unknown
 	cursor := err
 	for cursor != nil {
-		cursor = errors.Unwrap(cursor)
-		if res, ok := status.FromError(cursor); ok && res != nil {
-			return cursor //nolint:wrapcheck
+		if res, ok := status.FromError(cursor); ok {
+			code = res.Code()
+			break
 		}
+		cursor = errors.Unwrap(cursor)
 	}
-	return err
+	return &gRPCStatusError{
+		code: code,
+		err:  err,
```

Comment:
The sticky bit here is that we end up with a somewhat verbose error message.  For example, if we end up with an error chain like:

```go
fmt.Errorf("wrap: %w", status.Errorf(codes.FailedPrecondition, "i personally do not like the montage pipeline"))
```
Then this now prints as: `wrap: rpc error: code = FailedPrecondition desc = i personally do not like the montage pipeline`, whereas before it was `i personally do not like the montage pipeline`.  What we really want, but is hard to get, is `wrap: i personally do not like the montage pipeline`.

I think what you have here is an improvement, since information is never lost, but there is some confusion cost for people looking at the messages.  (The hard-coded string "rpc error" that status.Error() adds might technically be inaccurate, for example.)  I don't think we can really do better, easily, because we can't get at the constructor (or private fields) of every error that implements `interface { GRPCStatus() *Status }`, so we can't rewrite them.  We also can't guarantee that we control the `Error` method on the chain of errors; it can format however it wants.  

Long term, we probably need our own API for this, so methods can annotate their replies with GRPC status codes, without introducing the "rpc error: code = %s desc = %s" in the middle of an error chain.  Sigh!

Reply:
But you know what, we rarely use `status.Error` or `status.Errorf` directly, so this all might be perfectly OK.  Just some stuff in identity that might look messy if we don't make a dedicated error type for them:

```
src/identity/identity.go
12:     ErrInvalidID = status.Error(codes.Internal, "ID does not exist")
15:     ErrAlreadyExists = status.Error(codes.Internal, "ID already exists")

src/auth/auth.go
102:    ErrNotActivated = status.Error(codes.Unimplemented, "the auth service is not activated")
106:    ErrAlreadyActivated = status.Error(codes.Unimplemented, "the auth service is already activated")
112:    ErrNotSignedIn = status.Error(codes.Unauthenticated, "no authentication token (try logging in)")
116:    ErrNoMetadata = status.Error(codes.Internal, "no authentication metadata (try logging in)")
120:    ErrBadToken = status.Error(codes.Unauthenticated, "provided auth token is corrupted or has expired (try logging in again)")
124:    ErrExpiredToken = status.Error(codes.Internal, "token expiration is in the past")

src/license/license.go
12:     ErrDuplicateClusterID = status.Error(codes.Unimplemented, "cluster ID is not unique")
15:     ErrInvalidIDOrSecret = status.Error(codes.Unimplemented, "cluster ID or secret is not valid")
18:     ErrNotActivated = status.Error(codes.Unimplemented, "cluster does not have enterprise features enabled")
```

Reply:
As discussed elsewhere, I think this is handled now - we pass through the auth cases properly, and the pattern in other parts of the code of implementing the status code interface will work properly

Reply:
Agree.